model : add LLADA 2.0 diffusion support

wsbagnsv1 opened the pull request:
I recently created feature request [#16973](https://github.com/ggml-org/llama.cpp/issues/16973), but decided to implement it myself.

**Status: Work in Progress (WIP)**

This is my first PR here. I aimed for backwards compatibility, but the implementation currently contains some temporary workarounds (such as the `ubatch` one in the diffusion CLI) which I will remove once the underlying issues are resolved.

**Current State:**
* **Success:** Inference seems to work generally well and the output is coherent.
* **Known Issues:** I am currently tracking down bugs regarding unexpected tokens and an issue with max length.

Any feedback or assistance would be appreciated!

wsbagnsv1 committed:
Implement LLaDA2MoeModel conversion in convert_hf_to_gguf.py

wsbagnsv1 committed:
Add LLADA2 architecture to constants

wsbagnsv1 committed:
Implement LLaDA2.0 support to diffusion-cli.cpp

Added LLaDA2.0 support

wsbagnsv1 committed:
Add llada2.cpp to CMakeLists.txt

wsbagnsv1 committed:
Add LLADA2 architecture support

wsbagnsv1 committed:
Add LLM_ARCH_LLADA2 to architecture list

wsbagnsv1 committed:
Add llada2.0 to llama-model.cpp

wsbagnsv1 committed:
Create llada2.cpp

wsbagnsv1 committed:
Add llm_build_llada2 struct to models.h

wsbagnsv1 committed:
Merge branch 'ggml-org:master' into master

wsbagnsv1 commented:
This is the output with this prompt "what is the meaning of life?", I tested it on cuda and cpu and both work generally fine. 

<img width="1901" height="655" alt="Capture" src="https://github.com/user-attachments/assets/6d0bd234-d418-4f45-a2a6-e08df2d1b862" />

wsbagnsv1 committed:
Merge branch 'ggml-org:master' into master

am17an commented on examples/diffusion/diffusion-cli.cpp:
```diff
@@ -277,35 +291,69 @@ static void diffusion_generate(llama_context *          ctx,
+        
+        if (params.is_llada2) {
+            // LLaDA2.0: blocks start from position 0
+            block_start = (params.schedule == BLOCK_BASED) ? block_num * params.block_length : 0;
```
It is fine to start from n_input as usual here to, the n_input part never gets touched so it's effectively a no-op

wsbagnsv1 replied:
you are right, ive updated it, next step is to try and get the llada2.0 specific code out of diffusion cli (;

am17an reviewed (commented):
I didn't review all of this. But preferably there should be no LLaDA 2.0 specific stuff in diffusion-cli, it should all be covered by diffusion-cli parameters and maybe gguf parameters. Since it already supports Dream and LLaDA 1.0 without any specific case handling, it should be possible to do this for LLaDA 2.0 as well. As such I don't find it the sampling to be very different from block-based sampling used in LLaDA 1.0

wsbagnsv1 committed:
Add proper fall-through for llada2.0

wsbagnsv1 committed:
Cleanup 1

wsbagnsv1 committed:
Cleanup 2

wsbagnsv1 committed:
Add EOS, Threshold and batch strategy

wsbagnsv1 committed:
Add parameters to conversion script 

Added diffusion parameters for LLaDA2.0 in GGUF writer.

wsbagnsv1 committed:
Cleanup3

wsbagnsv1 committed:
Remove LLaDA2.0 specific code and make it model independent

wsbagnsv1 committed:
small fix

wsbagnsv1 committed:
small fix part 2

am17an commented on examples/diffusion/diffusion-cli.cpp:
```diff
@@ -622,6 +735,31 @@ int main(int argc, char ** argv) {
+        // Default to false for backward compatibility
+        diff_params.truncate_batch = false;
+    }  
+        
```
honestly these can just be diffusion-cli args, it should be possible to use this kind of sampling with other models as well

wsbagnsv1 replied:
although im not 100% certain we should do only cli args, since random users probably dont know about that, so it might be better to use them from the gguf directly?

wsbagnsv1 replied:
No idea what would be better tbh 😅

am17an replied:
threshold and eos_early_stop should both be diffusion-cli parameters. TBH I think truncate should be the default batch strategy for block-based sampling, but I will have to confirm

wsbagnsv1 committed:
small fix 1

wsbagnsv1 committed:
Merge branch 'master' into master

wsbagnsv1 commented:
The performance improvement increases time per step from 19.47ms -> 4.28ms for a 700ish token output, the longer the generation/context the better the performance improvement!

am17an commented on examples/diffusion/diffusion-cli.cpp:
```diff
@@ -267,6 +275,7 @@ static void diffusion_generate(llama_context *          ctx,
     if (params.schedule == BLOCK_BASED) {
         GGML_ASSERT(params.max_length % params.block_length == 0);
         num_blocks = params.max_length / params.block_length;
+       
```
```suggestion
```

am17an commented on examples/diffusion/diffusion-cli.cpp:
```diff
@@ -638,7 +776,7 @@ int main(int argc, char ** argv) {
     diff_params.steps            = params.diffusion.steps;
     diff_params.algorithm        = static_cast<diffusion_algorithm>(params.diffusion.algorithm);
-    diff_params.max_length       = params.n_ubatch;
+    diff_params.max_length       = max_length;
```
this need to be ubatch

am17an commented on examples/diffusion/diffusion-cli.cpp:
```diff
@@ -457,42 +481,124 @@ static void diffusion_generate(llama_context *          ctx,
+                            if (item.first > threshold) {
+                                high_conf_count++;
+                            }
                         }
```
Looking at https://huggingface.co/inclusionAI/LLaDA2.0-mini/blob/main/modeling_llada2_moe.py, I think this needs to be after the sampling

am17an commented on examples/diffusion/diffusion-cli.cpp:
```diff
@@ -277,7 +286,8 @@ static void diffusion_generate(llama_context *          ctx,
 
-    for (int block_num = 0; block_num < num_blocks; block_num++) {
+    bool all_tokens_filled = false;
+    for (int block_num = 0; block_num < num_blocks && !all_tokens_filled; block_num++) {
```
I don't think this is required here as we break from the loop when this eos token appears

wsbagnsv1 replied:
Im not so sure here, since it could be the case that the model generates an eos token in the diffusion process without being finished with the block which would kill this if I remember correctly, im gonna check anyways though

wsbagnsv1 commented:
Okay im not sure how we could add this without gating it by a diffusion parameter in the gguf, so i think ill just do that, if anyone has a better idea feel free to elaborate (;

wsbagnsv1 committed:
Enable hybrid diffusion

wsbagnsv1 committed:
Add HYBRID_DIFFUSION constant to diffusion class

wsbagnsv1 committed:
Remove LLM_ARCH_LLADA2 from architecture switch

Removed LLM_ARCH_LLADA2 case from switch statement.

wsbagnsv1 committed:
Implement hybrid diffusion optimization

Added hybrid diffusion optimization support to the diffusion parameters and processing logic.

wsbagnsv1 committed:
Make model use kv cache

wsbagnsv1 committed:
Add Hybrid diffusion mechanism 

This allows the model to use kv cache outside of the diffusion block speeding the model up when context builds up

wsbagnsv1 committed:
Clear white space

Co-authored-by: Aman Gupta <amangupta052@gmail.com>

wsbagnsv1 committed:
revert ubatch

wsbagnsv1 committed:
Change log level from INFO to DEBUG

wsbagnsv1 committed:
Improve confidence handling 

Refactor confidence calculation and transfer logic for clarity and efficiency.

wsbagnsv1 committed:
Refactor confidence calculation and transfer logic for clarity and efficiency.

CISC commented on convert_hf_to_gguf.py:
```diff
@@ -8729,6 +8729,46 @@ def prepare_tensors(self):
+        # Use the same tensor mapping as BailingMoeV2Model since they have identical tensor structure
+        self.tensor_map = gguf.get_tensor_name_map(self.model_arch, self.block_count)
+
+    def set_gguf_parameters(self):
```
```suggestion
    def set_gguf_parameters(self):
```
This serves no purpose.

CISC commented on convert_hf_to_gguf.py:
```diff
@@ -8729,6 +8729,46 @@ def prepare_tensors(self):
+        self.gguf_writer.add_string(gguf.Keys.Diffusion.BATCH_STRATEGY, "truncate")
+        self.gguf_writer.add_bool(gguf.Keys.Diffusion.EOS_EARLY_STOP, True)
+        self.gguf_writer.add_float32(gguf.Keys.Diffusion.THRESHOLD, 0.95)
+        self.gguf_writer.add_bool(gguf.Keys.Diffusion.HYBRID_DIFFUSION, True)
```
```suggestion
```
This can all go as well.
* Never use `add_key_value` unless you have to, and these are unused, so off they go.
* `pad_token_id` gets set by the vocab code.
* The diffusion parameters can be commandline options as suggested by @am17an

CISC commented on convert_hf_to_gguf.py:
```diff
@@ -8729,6 +8729,46 @@ def prepare_tensors(self):
+        
+    def set_vocab(self):
+        # Use the same vocab as BailingMoeV2Model
+        self._set_vocab_gpt2()
```
```suggestion
        self.gguf_writer.add_diffusion_shift_logits(False)
```
Already inherited.

wsbagnsv1 committed:
Implement EOS token assertion for early stop

Add assertion for EOS token early stopping.

CISC commented on src/models/llada2.cpp:
```diff
@@ -0,0 +1,133 @@
+    ggml_tensor * inp_out_ids = build_inp_out_ids();
+
+    const int n_transformer_layers = n_layer - hparams.nextn_predict_layers;
+    for (int il = 0; il < n_transformer_layers; ++il) {
```
```suggestion
    for (int il = 0; il < n_layer; ++il) {
```

wsbagnsv1 committed:
Merge branch 'ggml-org:master' into master

wsbagnsv1 committed:
Update src/models/llada2.cpp

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

wsbagnsv1 committed:
Update src/models/llada2.cpp

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

wsbagnsv1 commented on examples/diffusion/diffusion-cli.cpp:
```diff
@@ -305,12 +338,63 @@ static void diffusion_generate(llama_context *          ctx,
+            int32_t batch_start_pos;
+            
+            // Hybrid Diffusion: Commit previous block to KV cache
+            if (params.hybrid_diffusion && block_num > 0 && step == 0) {
```
Hmm, i think there was an issue when not using truncate for some reason the output becomes gibberish? Like with the --diffusion-hybrid flag (which i now use to toggle the truncate as well) it works fine but without it it stops with a "ΓÇ£" token 🤔

am17an replied:
I'm not entirely sure what you're saying, but you can ask a question in Q&A as I'm not too familiar with the kv cache. Nevertheless, your original implementation without any "hybrid diffusion" stuff was closer to being in a state where it could be merged. So my suggestion to you would be to get that in first, before tackling optimisations.

wsbagnsv1 committed:
Update convert_hf_to_gguf.py

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

wsbagnsv1 committed:
Update gguf-py/gguf/constants.py

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

wsbagnsv1 committed:
Update convert_hf_to_gguf.py

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

wsbagnsv1 committed:
Update convert_hf_to_gguf.py

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

wsbagnsv1 committed:
Update src/llama-arch.cpp

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

CISC commented on gguf-py/gguf/constants.py:
```diff
@@ -317,6 +317,10 @@ class Projector:
+        BATCH_STRATEGY      = "diffusion.batch_strategy"
+        EOS_EARLY_STOP      = "diffusion.eos_early_stop"
+        THRESHOLD           = "diffusion.threshold"
+        HYBRID_DIFFUSION    = "diffusion.hybrid_diffusion"
```
```suggestion
```

wsbagnsv1 committed:
Update gguf-py/gguf/constants.py

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

wsbagnsv1 committed:
Refactor EOS and threshold parameters to use CLI

wsbagnsv1 committed:
Add threshold and early stop flags to common.h

wsbagnsv1 committed:
Add diffusion options for threshold and early stopping

wsbagnsv1 committed:
Add options for hybrid diffusion

wsbagnsv1 committed:
Add hybrid diffusion optimization flag

wsbagnsv1 committed:
Remove truncate_batch and simplify hybrid diffusion

coder543 commented:
LLaDA 2.2 flash has been released now: https://huggingface.co/inclusionAI/LLaDA2.2-flash

IngwiePhoenix commented:
Has this PR gone stale?
